README.txt
currency.converter Package Readme
==========================================

Overview
-----------
currency.converter package fetches currency rate data from European Central Bank for about recent 90 days. Once the data is fetched the data is conserved in ZODB in case of whatever difficulty to fetch the data again. There are currency converter page and portlet included in this package and several methods you can use for your own applications.

This package is developed for plone-3.x.

The easiest way to demonstrate this package, just describe corrency.converter to the buildout.cfg to the egg section like below.

eggs =
    currency.converter

Log in as a manager and go to the next page "your_portal/@@manage-currency".
By visiting the page, the current currency data will be fetched if possible and automatically persisted.

-----------
Features
-----------
Site manager can give two different variables in addition to currencies, days and margin.

-Days
This amout is used to caclulate average of currencies. For example, if you input 10 to this field, 10 recent days are used to calculate average currency rate. This keeps currency rate fluctuation smaller than using everyday plain rate. If nothing or 0 is input there, it doesn't calculate average, but uses current rate.

-Margin
Margin adds % of rate to the currency rate. 0 is 0 % margin where is no margin.

-------------------------
Setting Time Server
-------------------------
To fetch the currency data every day, describe the next to the instance section of buildout.cfg.

[instance]
zope-conf-additional =
    <clock-server>
      method /your_portal/@@manage-currency
      period 86400
      user admin
      password admin_pass
      host localhost
    </clock-server>

* Change your_portal to your plone site id
* Period is seconds between eache fetch of the currency data. 86400 seconds are the 24 hour. The data is usually updated daily on week days.
* Describe admin name to admin and its password to admin_pass.
* host is host name of your server.

--------------------------------
Upgrading of reinstalling
--------------------------------
When you reinstall or upgrade to new version, the persisted data is not migrated to new environment, so please visit your_portal/@@manage-currency to get the current data.

------
ToDo
------
-Implement daily fetch of currency data automatically by using zope time server. Must be easy, but not tested yet.



docs/HISTORY.txt
Changelog
===========

0.5.5 (2010-08-25)
------------------
* Path to sample-90d.xml fixed for test.

0.5.4 (2010-08-24)
------------------
* Added IPortalCurrency to adapt currency to portal.
* Tested with Plone-4.0rc1.

0.5.3 (2010-06-03)
------------------
* currency-converter view updated.
* Deprecated import updated.
* Tested with Plone-4.0b3.

0.5.2 (2010-05-02)
------------------
* Minor visual updates.
* Tested with Plone-4.0b2.

0.5.1 (2009-08-27)
------------------
* Polish currency name changed to avoid UnicodeEncodeError.
* Tested with Plone-3.3.

0.5.0 (2009-08-13)
------------------
* Sample xml file added for test.
* Tested with Plone-3.3rc5.

0.4.4 (2009-06-18)
------------------
* Translations updated.

0.4.2 (2009-06-09)
------------------
* Currency code and name vocabulary added.

0.3.6 (2009-01-18)
------------------
* Initial release


currency/converter/tests/test_setup.py
import unittest
from Products.CMFCore.utils import getToolByName
from currency.converter.tests.base import CurrencyConverterTestCase
from currency.converter.interfaces import ICurrencyData
from zope.component import queryUtility

class TestSetup(CurrencyConverterTestCase):

    def afterSetUp(self):
        self.types = getToolByName(self.portal, 'portal_types')
        self.properties = getToolByName(self.portal, 'portal_properties')

    ## componentregistry.xml
    def test_utility(self):
        utility = queryUtility(ICurrencyData)
        self.assertNotEquals(None, utility)

    ## Propertiestool.xml
    def test_currency_xml(self):
        self.assertEquals('http://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist-90d.xml', self.properties.currency_converter_properties.getProperty('currency_xml'))

    def test_base_currency(self):
        self.assertEquals('EUR', self.properties.currency_converter_properties.getProperty('base_currency'))

    def test_days_for_avarage(self):
        self.assertEquals(1, self.properties.currency_converter_properties.getProperty('days_for_avarage'))

    def test_margin_for_avarage(self):
        self.assertEquals(0, self.properties.currency_converter_properties.getProperty('margin_for_avarage'))


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(TestSetup))
    return suite


currency/converter/tests/test_doctest.py
import unittest
import doctest
from zope.testing import doctestunit
from zope.component import testing
from Testing import ZopeTestCase as ztc
from currency.converter.tests import base

class TestSetup(base.CurrencyConverterFunctionalTestCase):

    def afterSetUp( self ):
        """Code that is needed is the afterSetUp of both test cases.
        """

        # Set up sessioning objects
        ztc.utils.setupCoreSessions(self.app)

def test_suite():
    return unittest.TestSuite([
        # Demonstrate the main content types
        ztc.FunctionalDocFileSuite(
            'tests/functional.txt', package='currency.converter',
            test_class=TestSetup,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE | doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS),

#        # Integration tests that use PloneTestCase
        ztc.ZopeDocFileSuite(
            'tests/integration.txt', package='currency.converter',
            test_class=base.CurrencyConverterTestCase,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE | doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS),

        # Unit tests
        doctestunit.DocFileSuite(
            'tests/unittest.txt', package='currency.converter',
            setUp=testing.setUp, tearDown=testing.tearDown,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE | doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS),

            ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


currency/converter/tests/test_portlet_currency_converter.py
from zope.component import getUtility, getMultiAdapter

from plone.portlets.interfaces import IPortletType
from plone.portlets.interfaces import IPortletManager
from plone.portlets.interfaces import IPortletAssignment
from plone.portlets.interfaces import IPortletDataProvider
from plone.portlets.interfaces import IPortletRenderer

from plone.app.portlets.storage import PortletAssignmentMapping

from currency.converter.portlets import currency_converter

from currency.converter.tests.base import CurrencyConverterTestCase

class TestPortlet(CurrencyConverterTestCase):

    def afterSetUp(self):
        self.setRoles(('Manager',))

    def testPortletTypeRegistered(self):
        portlet = getUtility(IPortletType, name='currency_converter.portlet')
        self.assertEquals(portlet.addview, 'currency_converter.portlet')

    def testInterfaces(self):
        portlet = currency_converter.Assignment()
        self.failUnless(IPortletAssignment.providedBy(portlet))
        self.failUnless(IPortletDataProvider.providedBy(portlet.data))

#    def testInvokeAddview(self):
#        portlet = getUtility(IPortletType, name='currency_converter.portlet')
#        mapping = self.portal.restrictedTraverse('++contextportlets++plone.leftcolumn')
#        for m in mapping.keys():
#            del mapping[m]
#        addview = mapping.restrictedTraverse('+/' + portlet.addview)

#        addview.createAndAdd(data={})

#        self.assertEquals(len(mapping), 1)
#        self.failUnless(isinstance(mapping.values()[0], currency_converter.Assignment))

#    def testInvokeEditView(self):
#        mapping = PortletAssignmentMapping()
#        request = self.folder.REQUEST

#        mapping['foo'] = currency_converter.Assignment()
#        editview = getMultiAdapter((mapping['foo'], request), name='edit')
#        self.failUnless(isinstance(editview, currency_converter.EditForm))

    def testRenderer(self):
        context = self.folder
        request = self.folder.REQUEST
        view = self.folder.restrictedTraverse('@@plone')
        manager = getUtility(IPortletManager, name='plone.rightcolumn', context=self.portal)
        assignment = currency_converter.Assignment()

        renderer = getMultiAdapter((context, request, view, manager, assignment), IPortletRenderer)
        self.failUnless(isinstance(renderer, currency_converter.Renderer))

#class TestRenderer(CurrencyConverterTestCase):
#    
#    def afterSetUp(self):
#        self.setRoles(('Manager',))
#        self.portal.invokeFactory('Cinema Folder', 'cf1')
#        self.portal.invokeFactory('Cinema Folder', 'cf2')
#        self.portal.cf1.invokeFactory('Promotion', 'p1')
#        self.portal.cf1.invokeFactory('Promotion', 'p2')
#        self.portal.cf1.invokeFactory('Promotion', 'p3')
#        self.portal.cf1.invokeFactory('Promotion', 'p4')
#        self.portal.cf1.invokeFactory('Promotion', 'p5')
#        self.portal.cf2.invokeFactory('Promotion', 'p6')
#        self.portal.cf2.invokeFactory('Promotion', 'p7')

#    def renderer(self, context=None, request=None, view=None, manager=None, assignment=None):
#        context = context or self.folder
#        request = request or self.folder.REQUEST
#        view = view or self.folder.restrictedTraverse('@@plone')
#        manager = manager or getUtility(IPortletManager, name='plone.rightcolumn', context=self.portal)
#        assignment = assignment or currency_converter.Assignment()

#        return getMultiAdapter((context, request, view, manager, assignment), IPortletRenderer)

#    def test_count(self):
#        r = self.renderer(context=self.portal.cf1, assignment=currency_converter.Assignment(count=5))
#        self.assertEquals(5, len([p for p in r.promotions()]))

#    def test_randomize(self):
#        r = self.renderer(context=self.portal.cf1, assignment=currency_converter.Assignment(count=5, randomize=True))
#        self.assertEquals(5, len([p for p in r.promotions()]))
#        # Mmmm, hard to test for random things :)
#        
#    def test_sitewide(self):
#        r = self.renderer(context=self.portal.cf1, assignment=currency_converter.Assignment(count=10, sitewide=True))
#        p6_url = self.portal.cf2.p6.absolute_url()
#        self.failUnless(p6_url in [p['url'] for p in r.promotions()])

def test_suite():
    from unittest import TestSuite, makeSuite
    suite = TestSuite()
    suite.addTest(makeSuite(TestPortlet))
#    suite.addTest(makeSuite(TestRenderer))
    return suite


currency/converter/interfaces.py
#from zope.interface import Attribute
from zope.interface import Interface
#from zope.schema import Container

from currency.converter import CurrencyConverterMessageFactory as _

class ICurrencyData(Interface):
    """CurrencyData itself: for methods which return currency rate, they return currency code as well. (currency_code, currency_rate) tuple?"""

    def currency_data():
        """Returns the most recent currency data."""

    def currency_data_list():
        """Returns the most recent currency data as a list of dictionaries.
        ex)
        [
            {
            code: "EUR",
            rate: [156.05, 160.25, ...],
            name: u"Euro",
            unit: u"\u20ac"
            },
            {...},{...},...
        ]
        """

    def updated_date():
        """Returns updated date."""

    def currency_codes():
        """Retrurns currency codes."""

    def currency_code_tuples():
        """Returns currency code and its tuples."""

    def currency_code_data():
        """Returns dictionary of currency code and rate list."""

    def currency_code_average(days):
        """Returns code and average rate of days."""

    def days():
        """Returns maximum gotten days."""

    def currency_rate_against_base_code(days, code):
        """Returns currency rate gainst base code."""

    def currency_rate_against_base_code_with_margin(days, code, margin):
        """Returns currency rate gainst base code with margin."""

    def currency_rate(days, margin, base_currency_code, base_rate, currency_code, currency_rate):
        """Returns calculated currency rate."""


currency/converter/browser/currency_manager.py
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
from Acquisition import aq_inner
#from Products.CMFCore.utils import getToolByName
from currency.converter import CurrencyConverterMessageFactory as _
#from zope.component import getMultiAdapter
from zope.component import queryUtility
from currency.converter.interfaces import ICurrencyData

class CurrencyManagerView(BrowserView):
    """View for currency manager."""

    template = ViewPageTemplateFile('templates/currency_manager.pt')

    def __call__(self):

#        ## Hide the editable-object border.
        self.request.set('disable_border', True)

#        ## Defines.
        form = self.request.form
        currency_data = queryUtility(ICurrencyData)
        if currency_data.currency_data():
            currency_data.currencies = currency_data.currency_data()
        self.updated_date = currency_data.updated_date()
        self.currency_code_tuples = currency_data.currency_code_tuples()
        self.days = currency_data.days()
        self.selected_currency = currency_data.selected_base_currency
        self.selected_days = currency_data.selected_days
        self.margin = currency_data.margin
        self.currencies = currency_data.currency_rate_against_base_code_with_margin(int(self.selected_days), self.selected_currency, self.margin)
        if self.currencies != False:
            self.currency_rate = self.currencies.items()
        else:
            self.currency_rate = False
        self.error_message = False

        ## Check buttons.
        update_button = form.get('form.button.Update', None) is not None

        if update_button:
            try:
                float(form.get('margin'))
                self.margin = float(form.get('margin'))
                currency_data.margin = self.margin
            except:
                self.error_message = _(u"Please input float like 5.00")
            self.selected_currency = form.get('currency_code')
            currency_data.selected_base_currency = self.selected_currency
            self.selected_days = int(form.get('days'))
            currency_data.selected_days = self.selected_days
            self.currencies = currency_data.currency_rate_against_base_code_with_margin(self.selected_days, self.selected_currency, self.margin)
            self.currency_rate = self.currencies.items()
            return self.template()

        else:
            return self.template()


currency/converter/browser/currency_converter.py
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

from Acquisition import aq_inner

from Products.CMFCore.utils import getToolByName

from currency.converter import CurrencyConverterMessageFactory as _

#from zope.component import getMultiAdapter
from zope.component import getUtility

from currency.converter.interfaces import ICurrencyData

class CurrencyConverterView(BrowserView):
    """View for currency manager."""

    template = ViewPageTemplateFile('templates/currency_converter.pt')

    def __call__(self):

#        ## Hide the editable-object border.
        self.request.set('disable_border', True)

#        ## Defines.
        form = self.request.form

        ## Data from ICurrencyData
        currency_data = getUtility(ICurrencyData)
        self.currency_data = currency_data.currencies
        self.updated_date = currency_data.updated_date()
        self.currency_code_tuples = currency_data.currency_code_tuples()
        self.days = currency_data.days()
        self.selected_base_currency = currency_data.selected_base_currency
        self.selected_days = currency_data.selected_days
        self.margin = currency_data.margin
#        self.currencies = currency_data.currency_rate_against_base_code_with_margin(int(self.selected_days), self.selected_currency, self.margin)

        try:
            self.base_currency_rate = form.get('base_currency_rate', 1)
        except:
            self.base_currency_rate = 1
#        try:
#            self.base_currency = form.get('base_currency', 'EUR')
#        except:
#            self.base_currency = 'EUR'
        try:
            self.selected_base_currency_code = form.get('base_currency_code', self.selected_base_currency)
        except:
            self.selected_base_currency_code = self.selected_base_currency
        try:
            self.selected_currency_code = form.get('currency_code', None)
        except:
            self.selected_currency_code = None


        self.calculated_rate = None

        self.error_message = None

        ## Check buttons.
        convert_button = form.get('form.button.Convert', None) is not None

        if convert_button:
            if form.get('base_currency_code') != form.get('currency_code'):
                self.base_currency_rate = form.get('base_currency_rate')
                self.base_currency_code = form.get('base_currency_code')
                self.currency_code = form.get('currency_code')
                self.calculated_rate = self.calculated_rate_against_base_rate(float(self.base_currency_rate), self.base_currency_code, self.currency_code)
                return self.template()
            else:
                self.error_message =_(u"Please choose different currencies.")
                return self.template()

        else:
            return self.template()

    def calculated_rate_against_base_rate(self, base_currency_rate, base_currency_code, currency_code):
        """Returns calculated rate against base currency rate."""
        currency_data = getUtility(ICurrencyData)
        days = currency_data.selected_days
        margin = currency_data.margin
        currency_dictionary = currency_data.currency_rate_against_base_code_with_margin(days, base_currency_code, margin)
        result = currency_dictionary[currency_code] * base_currency_rate
        return '%.2f' %result
#        return currency_dictionary[currency_code]
#        return currency_dictionary["USD"]


currency/converter/currency_data.py
import warnings
from persistent import Persistent
from zope.interface import implements
from currency.converter.interfaces import (
                                            ICurrencyData,
                                            IRateAgainstBaseRate,
                                            ICurrencyCodeName,
                                            ICurrencyCodeNameTuples,
                                            )
from elementtree.ElementTree import ElementTree#, XML
import urllib2
from currencies import currencies

from zope.component import getUtility

class CurrencyData(Persistent):
    implements(ICurrencyData)

    def __init__(self):
        self.currencies = None
        self.selected_base_currency = "EUR"
        self.selected_days = 1
        self.margin = 0.00
        self.date = None
        self.codes = ['EUR']

    def currency_data(self, xml='http://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist-90d.xml'):
        """Returns the most recent currency data with tuples."""
        etree90 = ElementTree()
        try:
            data90 = urllib2.urlopen(xml)
            root90 = ElementTree.parse(etree90, data90)
            DATA90 = root90[2]
            DATA_list = []
            for DATA in DATA90:
                daily_data_list = []
                for daily_data in DATA:
                    daily_data_tuple = (daily_data.get('currency'), daily_data.get('rate'))
                    daily_data_list.append(daily_data_tuple)
                ddl = (DATA.get('time'), dict(daily_data_list))
                DATA_list.append(ddl)
                date = DATA_list[0][0]
#            if (
#                (self.currencies is not None and self.date != date) or
#                (self.currencies is None)
#            ):
            if self.date != date:
                self.currencies = DATA_list
                self.date = date
                for date in self.currencies:
                    for key in date[1].keys():
                        if key not in self.codes:
                            self.codes.append(key)
        except:
            pass

    def currency_data_list(self):
        """Returns the most recent currency data as a list of dictionaries.
        ex)
        [
            {
            code: "EUR",
            rate: ['156.05', '160.25', ...],
            name: u"Euro",
            unit: u"\u20ac"
            },
            {...},{...},...
        ]
        """
        if self.currencies is not None:
            daily_data_list = [daily_data[1] for daily_data in self.currencies]
            results = []
            for code in self.currency_codes():
                rate_list = [d.get(code) for d in daily_data_list if d.get(code) != None]
                try:
                    if code == "EUR":
                        results.append(
                    {
                    'code': code, 
                    'rate': [1],
                    'name': currencies.get(code)[0],
                    'unit': currencies.get(code)[1],
                    'decimal' : currencies.get(code)[2],
                    }
                    )
                    else:
                        results.append(
                    {
                    'code': code, 
                    'rate': rate_list,
                    'name': currencies.get(code)[0],
                    'unit': currencies.get(code)[1],
                    'decimal' : currencies.get(code)[2],
                    }
                    )
                except:
                    results.append(
                    {
                    'code': code, 
                    'rate': rate_list,
                    'name': currencies.get(code)[0],
                    'unit': currencies.get(code)[1],
                    'decimal' : 2,
                    }
                    )
            return results
        else:
            return False

    def updated_date(self):
        """Returns updated date."""
        warnings.warn(
            ("'update_date' is deprecated and will be removed in version 0.6"),
            DeprecationWarning
        )
        if self.currencies != None:
            return self.currencies[0][0]
        else:
            return False

    def currency_codes(self):
        """Retrurns currency codes."""
        warnings.warn(
            ("'update_date' is deprecated and will be removed in version 0.6"),
            DeprecationWarning
        )
        codes = ['EUR']
        if self.currencies is not None:
            for date in self.currencies:
                for key in date[1].keys():
                    if key not in codes:
                        codes.append(key)
        return codes

    def currency_code_tuples(self):
        """Returns currency code and its tuples."""
        results = []
        for code in self.currency_codes():
            t = (code, currencies[code])
            results.append(t)
        return results

    def currency_code_data(self):
        """Returns dictionary of currency code and rate list."""
        if self.currencies is not None:
            currency_data_withought_date = []
            for data in self.currencies:
                currency_data_withought_date.append(data[1])
            r = {}
            for code in self.currency_codes():
                results = []
                for data in currency_data_withought_date:
                    for key in data.keys():
                        if code == key:
                            results.append(data[key])
                r.update({code:results})
            return r
        else:
            return False

    def currency_code_average(self, days):
        """Returns code and average rate of days."""
        if self.currencies is not None:
            results = {}
            for (code, L) in self.currency_code_data().items():
                if code == 'EUR':
                    rate = 1
                    results.update({code:rate})
                elif len(L) >= days > 0:
                    L = L[0:days]
                    S = 0
                    for l in L:
                        S = S + float(l)
                    rate = S / days
                    results.update({code:rate})
                elif days > len(L) > 0:
                    S = 0
                    for l in L:
                        S = S + float(l)
                    rate = S / len(L)
                    results.update({code:rate})
                else:
                    pass
            return results
        else:
            return False

    def days(self):
        """Returns maximum gotten days."""
        if self.currencies is not None:
            return range(1,len(self.currencies)+1)
        else:
            return [1]

    def currency_rate_against_base_code(self, days, code):
        """Returns currency rate gainst base code."""
        if self.currencies != None:
            if code == "EUR":
                cca = self.currency_code_average(days)
                del cca[code]
                return cca
            else:
                cca = self.currency_code_average(days)
                code_value = float(cca[code])
                del cca[code]
                results = {}
                for (k, v) in cca.items():
                    v = float(v) / code_value
                    results.update({k:v})
                return results
        else:
            return False

    def currency_rate_against_base_code_with_margin(self, days, code, margin):
        """Returns currency rate gainst base code with margin."""
        if self.currencies != None:
            results = {}
            for (k, v) in self.currency_rate_against_base_code(days, code).items():
                v = v * (100 + margin) / 100
                results.update({k:v})
            return results
        else:
            return False

    def currency_rate(self, days=1, margin=0, base_currency_code="EUR", base_rate=1, currency_code="USD", currency_rate=10):
        """Returns calculated currency rate."""
        if base_currency_code == "EUR":
            for cd in self.currency_data_list():
                if cd['code'] == currency_code:
                    rates = cd['rate']
                    decimal = cd['decimal']
                    break


class RateAgainstBaseRate(object):
    """A component which provides rate agains base currency rate."""

    implements(IRateAgainstBaseRate)

    def __call__(self, base_currency_rate, base_currency_code, currency_code):
        """Returns currency rate for base currency rate, margin and days."""
        currency_data = getUtility(ICurrencyData)
        days = currency_data.selected_days
        margin = currency_data.margin
        currency_dictionary = currency_data.currency_rate_against_base_code_with_margin(days, base_currency_code, margin)
        if currency_code == None:
            return None
        if currency_code == '':
            return None
        elif base_currency_code != currency_code:
            result = currency_dictionary[currency_code] * base_currency_rate
            return '%.2f' %result
        else:
            return None

class CurrencyCodeName(object):
    """
    A component which provides list of dictionaries for currency code and name.
    """

    implements(ICurrencyCodeName)

    def __call__(self):
        """Returns list of dictionaries for currency code and name."""
        currency_data = getUtility(ICurrencyData)
        currency_data_list = currency_data.currency_data_list()
        return [{'code':i['code'],'name':i['name']} for i in currency_data_list]

class CurrencyCodeNameTuples(object):

    implements(ICurrencyCodeNameTuples)

    def __call__(self):
        """Returns tuple of tuples for currency code and name."""
        currency_data = getUtility(ICurrencyData)
        currency_data_list = currency_data.currency_data_list()
        l = [(i['code'],i['name']) for i in currency_data_list]
        t = tuple(l)
        return t
